### PR TITLE
Reject introspection requests whose client is not a known audience #156

### PR DESCRIPTION
**Pull request by MonkeyJamboree**
Implemented fix for https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/issues/156

### CODE REVIEW

File: src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs [hunk 1]
```diff
@@ -1782,6 +1782,17 @@ internal partial class OpenIdConnectServerHandler : AuthenticationHandler<OpenId
                 return;
             }
 
+            // Note: reject requests whose client is not a known audience.
+            if (notification.Audiences.Contains(clientNotification.ClientId)) {
```
Comment (PinpointTownes) [inline src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs]: Missing `!`? :smile: 

Note that this check won't work if the client was not authenticated.

Comment (MonkeyJamboree) [inline src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs]: Oops, yeah, forgot to negate it. Other than that I wasn't sure how else to determine the audience.

-----Original Message-----

**Comment by MonkeyJamboree** [comment]
After thinking about it, out of curiosity, how would the client not be authenticated? Shouldn't `Options.Provider.ValidateClientAuthentication()` check that the client is authenticated via client credentials? In what circumstances would the introspection endpoint be called without authentication?

**Comment by PinpointTownes** [comment]
We agreed not to make client authentication mandatory to support public clients (like JS or mobile apps), so if the developer calls `context.Skipped()` from the `ValidateClientAuthentication` event, the introspection/validation request is processed without `clientNotification.ClientId` being set.

**Comment by MonkeyJamboree** [comment]
I was reading some of the mailing lists, and according to [this thread](https://www.ietf.org/mail-archive/web/oauth/current/msg14783.html) it seems that the introspection endpoint is not designed for client consumption in any way, just for resource servers. In what sense would a client logistically be using the introspection endpoint instead of an identity token (for checking that an authentication occurred), gaining a new access token when authorization fails (for any reason) at a resource server, or using the userinfo endpoint for getting detailed information about a token? As stated in the mailing list, you likely wouldn't have a "public client resource server" out there.

**Comment by PinpointTownes** [comment]
The validation endpoint is heavily used by clients whose limited resources (or the lack of efficient JWT libraries) prevent them from validating identity tokens locally. It's also frequently used to determine whether an access token or a refresh token is still valid without having to query the resource server.

I agree that keeping the validation/introspection endpoint standard is important, but mandatory client authentication would make our validation stuff almost useless.

**Comment by PinpointTownes** [comment]
BTW, I'm not the only one having such concerns about this introspection endpoint :smile: 

> It wouldn't have to be an all-or-nothing policy choice either, a server could chose to reject requests from confidential clients where client authentication is not provided, but accept requests without client authentication from non-confidential clients.  A server that has sufficiently high entropy in the tokens, abuse protection on the endpoint, and is not concerned about an unrelated party (that happens to have a token intended for a different party) learning the token metadata, could simply not require any client authentication at all.

**Comment by MonkeyJamboree** [comment]
I would actually argue the point about determining if an access or refresh token is valid - if an access token is invalid when calling a resource server, you'll know directly based on the 401 you get back, and refresh tokens shouldn't ever be sent to resource servers, only to authorization servers to re-acquire a valid access token. If a client is first checking that an access token is valid via the introspection/validation endpoint, THEN making a call to a resource server, there is a potential for a time differential in which the access token would expire (causing a 401 anyway), in addition to the overhead of making a second, unnecessary, and potentially problematic HTTP call.

Yes, I saw the arguments from that perspective but I still don't understand the intent behind it. No replies were made when asked
> Are you trying to use token introspection as evidence of authentication of the user?

I could understand wanting to have an endpoint or API that does JWT validation for you if you can't validate on the client for some reason, but that seems to be a separate problem from what the introspection endpoint is there to solve. Not to mention if you're querying the authorization server that issued the JWT to validate it, that seems to bring in more possibility for Man-in-the-middle attacks of some kind... But if that needed to be done, why not use the userinfo endpoint since it's specifically designed for client consumption of tokens?

I'm not saying to make the introspection endpoint inflexible, I just want to be sure we're following the spec appropriately and not leaving things open for security risks if many other applications want to use the middleware. Perhaps an option for allowing unauthenticated introspection requests?

**Comment by MonkeyJamboree** [comment]
I think the key points in William's argument are the following:
- sufficiently high entropy in the tokens
- abuse protection on the endpoint
- allowance of any party to learn about any token's metadata

In my opinion, it should be up to consumers of the middleware and/or implementers of `IOpenIdConnectServerProvider` to explicitly allow this behavior if needed and account for it, but by default to follow the spec when implementing the logic.

Also, should these comments be moved to or referenced in a new Issue as a discussion topic?

**Comment by PinpointTownes** [comment]
> I would actually argue the point about determining if an access or refresh token is valid - if an access token is invalid when calling a resource server, you'll know directly based on the 401 you get back, and refresh tokens shouldn't ever be sent to resource servers, only to authorization servers to re-acquire a valid access token. If a client is first checking that an access token is valid via the introspection/validation endpoint, THEN making a call to a resource server, there is a potential for a time differential in which the access token would expire (causing a 401 anyway), in addition to the overhead of making a second, unnecessary, and potentially problematic HTTP call.

You assume a validation request is **always** made before an API call. Sadly, you're wrong :smile:
I recently help reviewing an app where the "token info" endpoint of a well known provider was used to determine whether the authorization (= the refresh token stored in the database) was still valid or not and to potentially remove the unused accounts.

> I could understand wanting to have an endpoint or API that does JWT validation for you if you can't validate on the client for some reason, but that seems to be a separate problem from what the introspection endpoint is there to solve. Not to mention if you're querying the authorization server that issued the JWT to validate it, that seems to bring in more possibility for Man-in-the-middle attacks of some kind...

Not sure why the validation endpoint would be more prone to MITM attacks than just a classic implicit roundtrip, so I'd be curious to see a practical case :+1: 

> But if that needed to be done, why not use the userinfo endpoint since it's specifically designed for client consumption of tokens?

The userinfo endpoint only accepts access tokens, not identity tokens. And it won't return critical data like the expiration date of the token. It clearly serves a different purpose.

> I'm not saying to make the introspection endpoint inflexible, I just want to be sure we're following the spec appropriately and not leaving things open for security risks if many other applications want to use the middleware. Perhaps an option for allowing unauthenticated introspection requests?

You can already block unauthenticated introspection/validation requests by calling `Rejected()` (which is the default behavior) from the `ValidateClientAuthentication` event. Not sure why a new option would be needed.

____________

If you really think our introspection/validation endpoint might be prone to token scanning attacks, feel free to submit a PR to ensure the custom tokens manually created by the end developers respect a given security level (minimal length, enough entropy). But I'm definitely convinced that requiring client authentication to avoid this kind of attacks is NOT the right answer, despite what the specification says (after all, nothing prevents an authenticated client from trying a token scanning attack :smile:)

**Comment by PinpointTownes** [comment]
> But I'm definitely convinced that requiring client authentication to avoid this kind of attacks is NOT the right answer, despite what the specification says (after all, nothing prevents an authenticated client from trying a token scanning attack :smile:)

Not to mention that a simple API/resource server can be targeted by a token scanning attack. None of the built-in middleware in aspnet/Security have specific countermeasures to avoid that.

**Comment by PinpointTownes** [comment]
Here's what we should implement:

Refresh token:

* If the caller is authenticated:
  * If the token was issued to an authenticated client AND the caller is the app corresponding to the client_id: return active/status/iss/aud/iat/exp.

  * If the token was not issued to a fully authenticated client OR if the caller is not the client app: return nothing except the active=false claim.

* If the caller is not authenticated: return active/status/iss/aud/iat/exp.


Identity token:

* If the caller is authenticated AND is one of the specified audiences (= confidential client app), return everything, even non-standard claims.

* If the caller is not authenticated OR is not one of the specified audiences (= public client app or unknown caller):
  * If the token was issued to an authenticated client: return nothing except the active=false claim.

  * If the token was not issued to a fully authenticated client: return active/status/iss/aud/iat/exp.


Access token:

* If the caller is authenticated AND is one of the specified audiences (= resource server), return everything, even non-standard claims.

* If the caller is not authenticated OR is not one of the specified audiences (= client app or unknown caller):
  * If the token was issued to an authenticated client:
    * If the caller is listed as an authorized party: return active/status/iss/aud/iat/exp.
    
    * If the caller is not listed as an authorized party: return nothing except the active=true claim.

  * If the token was not issued to a fully authenticated client: return active/status/iss/aud/iat/exp.

**Comment by MonkeyJamboree** [comment]
> You assume a validation request is always made before an API call. Sadly, you're wrong :smile:
I recently helped reviewing an app where the "token info" endpoint of a well known provider was used to determine whether the authorization (= the refresh token stored in the database) was still valid or not to potentially remove the unused accounts.
Actually, I was just thinking in general access token terms. I hadn't thought about refresh token validity in this case, but public (AKA non-confidential) clients in OAuth 2.0 shouldn't get refresh tokens to begin with since they can't keep them secure... 
The OAuth 2.0 spec says:
>Refresh tokens MUST be kept confidential in transit and storage, and shared only among the authorization server and the client to whom the refresh tokens were issued.
The key word there being confidential, meaning a client who CAN keep them confidential (in other words, NOT a public client). Using a server-side application in combination with the public client allows the usage of refresh tokens in some way as you may be able to store it in an encrypted cookie, associate a refresh token with a user/session of some kind, or use some other technique, but that still doesn't get around the issue that the public client *itself* should never be able to see the refresh token for security reasons, and you're inherently introducing a confidential client that delegates access to a public client anyway.
Also, In section [9. Client Authentication](http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) for the OIDC spec, it says:
> During Client Registration, the RP (Client) MAY register a Client Authentication method. If no method is registered, the default method is client_secret_basic.
and under "none":
> The Client does not authenticate itself at the Token Endpoint, either because it uses only the Implicit Flow (and so does not use the Token Endpoint) or because it is a Public Client with no Client Secret or other authentication mechanism.
To me, this means that since a public client can't authenticate, and can't keep refresh tokens confidential, the AS shouldn't issue them to the client... but I don't know enough about the OIDC spec to say if that's true. I'm sure if consumers really wanted to issue refresh tokens to public clients they could... Also, I'm seeing a lot in the OIDC spec about not necessarily requiring authentication when making requests to the `token` endpoint, so I feel like this is either grey area or potentially a conflict between the specs.
Another way to check if a refresh token is valid is to simply make a refresh token flow request and see if you get an access token, but I could understand why that's more than you really need, or if you need more detailed information about the token expiration itself and such...

Right now I'm trying to take a bit of a step back to understand all the various use-cases for all of this... Can you provide more details about the circumstances where apps you've helped with have had the need to call the validation endpoint (besides checking for refresh token validity)? In what circumstances would a client NOT be able to validate tokens? Also, in what circumstances would an unauthenticated client be able to get tokens, given that at least the `client_id` and `redirect_uri` are required at some points in the pipeline?

**Comment by PinpointTownes** [comment]
> I hadn't thought about refresh token validity in this case, but public (AKA non-confidential) clients in OAuth 2.0 shouldn't get refresh tokens to begin with since they can't keep them secure... 

Heh? Of course they can keep them "safe", just like any other cookie or value in the local storage. And in OAuth2, issuing a refresh token is a totally legal operation:

> If the client type is confidential or the client was issued client credentials (or assigned other authentication requirements), the client MUST authenticate with the authorization server as described in Section 3.2.1.

Which means that non-confidential apps like JS or mobile apps are free to use the refresh token grant without having to authenticate (https://tools.ietf.org/html/rfc6749#section-6).

In OIDC, client authentication is mandatory (http://openid.net/specs/openid-connect-core-1_0.html#RefreshingAccessToken) but we accept unauthenticated requests to keep ASOS OAuth2-compatible.

> The key word there being confidential, meaning a client who CAN keep them confidential (in other words, NOT a public client). Using a server-side application in combination with the public client allows the usage of refresh tokens in some way as you may be able to store it in an encrypted cookie, associate a refresh token with a user/session of some kind, or use some other technique, but that still doesn't get around the issue that the public client itself should never be able to see the refresh token for security reasons, and you're inherently introducing a confidential client that delegates access to a public client anyway.

As you've already figured out, keeping the access/refresh tokens in a "safe" container has **nothing** to do with client authentication. You're mixing up both things.

> Another way to check if a refresh token is valid is to simply make a refresh token flow request and see if you get an access token, but I could understand why that's more than you really need, or if you need more detailed information about the token expiration itself and such...

Yes, but a refresh token request can have unwanted side effects. The most common one is automatic token revocation: when a new token is obtained, the old ones can be automatically revoked, which is definitely not what you want when making a simple validation query.

> Can you provide more details about the circumstances where apps you've helped with have had the need to call the validation endpoint (besides checking for refresh token validity)? In what circumstances would a client NOT be able to validate tokens?

Recently, I participated to the elaboration of a UWP app. Sadly, IdentityModel is not compatible with this platform yet. Of course, re-creating a JWT library from scratch was NOT an option and we preferred using the validation endpoint.

> Also, in what circumstances would an unauthenticated client be able to get tokens, given that at least the client_id and redirect_uri are required at some points in the pipeline?

ROPC requires neither `client_id` (for public apps) nor `redirect_uri` :smile:

**Comment by MonkeyJamboree** [comment]
Huh. I never looked at the OAuth spec in that way, but I guess it really doesn't dictate that refresh tokens can't be issued to public clients... It does say this:

>The authorization server must consider the security implications of interacting with unauthenticated clients and take measures to limit the potential exposure of other credentials (e.g., refresh tokens) issued to such clients.

...but yeah, that doesn't mean they can't be issued.

> ROPC requires neither client_id (for public apps) nor redirect_uri :smile:

Um. Okay, I'm seeing the entire spec in an entirely new light today... Am I seeing it right that client authentication is really only required for the Client Credentials grant?

:open_mouth: :hear_no_evil: :boom: I think my head just asploded.

Thanks for all the feedback. I definitely need to mull this all over now!

**Comment by MonkeyJamboree** [comment]
I'm questioning some of the logic about what we should implement... with Refresh tokens, if the caller of the introspection endpoint was not authenticated, but the token was issued to an authenticated client, shouldn't the response actually be the active=false response since the AS needs to authenticate when a client is confidential? (To me this says the client didn't send their credentials or it's a different client trying to find out about a refresh token's validity.)

In the case of Access tokens, how do you determine if the caller is an authorized party? Is there something that sets this somewhere in the middleware? If the caller is not listed as an authorized party, shouldn't the claim be active=false (not true)?
Also, if the caller is authenticated, but isn't one of the specified audiences (or an authorized party of some kind), shouldn't active=false be returned? Are there circumstances where the caller wouldn't be either an audience or authorized party?

Here's what I see as the adjusted logic for these cases:
 - Refresh token (should only be known to the AS and the client for which it's issued, according to the OAuth 2.0 spec):
  - If the caller is authenticated (= confidential client app or resource server):
     - If the token was issued to an authenticated client AND the caller is the app corresponding to the client_id (= same client): return active/status/iss/aud/iat/exp.
     - If the token was not issued to an authenticated client OR if the caller is not the client app (= different caller): return nothing except the active=false claim.
  - If the caller is not authenticated (= public client app or unknown caller): 
     - If the token was issued to an authenticated client (= different client?): return nothing except the active=false claim.
     - If the token was not issued to an authenticated client (= public client app): return active/status/iss/aud/iat/exp.

 - Identity token:
  - If the caller is authenticated AND is one of the specified audiences (= confidential client app), return everything, even non-standard claims.
  - If the caller is not authenticated (= public client app or unknown caller) OR is not one of the specified audiences (different caller):
     - If the token was issued to an authenticated client: return nothing except the active=false claim.
     - If the token was not issued to a fully authenticated client: return active/status/iss/aud/iat/exp.

 - Access token:
  - If the caller is authenticated AND is one of the specified audiences (= resource server), return everything, even non-standard claims.
  - If the caller is not authenticated (= public client app or unknown caller):
     - If the token was issued to an authenticated client:
         - If the caller is listed as an authorized party (how do you determine this?): return active/status/iss/aud/iat/exp.
         - If the caller is not listed as an authorized party (how do you determine this?): return nothing except the active=false claim.
     - If the token was not issued to a fully authenticated client: return active/status/iss/aud/iat/exp.
   - If the caller is authenticated and is NOT one of the specified audiences or an authorized party (= invalid audience): return nothing except the active=false claim?

Thoughts?

**Comment by PinpointTownes** [comment]
> I'm questioning some of the logic about what we should implement... with Refresh tokens, if the caller of the introspection endpoint was not authenticated, but the token was issued to an authenticated client, shouldn't the response actually be the active=false response since the AS needs to authenticate when a client is confidential? (To me this says the client didn't send their credentials or it's a different client trying to find out about a refresh token's validity.)

Yeah, sure. My table was not explicit enough :+1: 

> In the case of Access tokens, how do you determine if the caller is an authorized party? Is there something that sets this somewhere in the middleware?

When the `client_id` is known, it is automatically added by `CreateAccessTokenAsync` as a `azp` claim (authorized party).

> If the caller is not listed as an authorized party, shouldn't the claim be active=false (not true)? Also, if the caller is authenticated, but isn't one of the specified audiences (or an authorized party of some kind), shouldn't active=false be returned?

Ah yeah, that's a typo :+1: 

> Are there circumstances where the caller wouldn't be either an audience or authorized party?

I guess we could find a specific scenario, but I'm not sure we should allow that.

**Comment by MonkeyJamboree** [comment]
Another question: If an Identity token is issued to a public client, but the caller is authenticated, should the introspection request return the active=false claim since it's a different caller asking about the token? Or is it possible to have authenticated clients asking about public clients' Identity tokens?

**Comment by PinpointTownes** [comment]
> Another question: If an Identity token is issued to a public client, but the caller is authenticated, should the introspection request return the active=false claim since it's a different caller asking about the token? Or is it possible to have authenticated clients asking about public clients' Identity tokens?

You can, but it wouldn't be a security feature, since the caller would just have to be unauthenticated to get more details :smile:

**Pull request closed without merging**